File: app/SankalpaCore/Sources/SankalpaCore/Domain/Lifecycle/LifecycleTimeline.swift
import Foundation

/// The transition history. It satisfies the audit requirement and is also the only thing that can
/// answer "was this sankalpa In progress when that session happened?" and "was this whole window
/// Paused?" (DD-4).
public struct LifecycleTimeline: Hashable, Codable, Sendable {
    public private(set) var current: LifecycleState
    public private(set) var transitions: [LifecycleTransition]

    public init() {
        self.current = .notStarted
        self.transitions = []
    }

    public init(current: LifecycleState, transitions: [LifecycleTransition]) {
        self.current = current
        self.transitions = transitions
    }

    // MARK: - Recording

    /// S7, S11, S14 — validates the transition table, then the timing rules, then appends.
    public mutating func record(
        to target: LifecycleState,
        effectiveAt: CalendarMoment,
        recordedAt: CalendarMoment,
        commitmentStart: CalendarDay
    ) throws(LifecycleTransitionError) {
        guard current.canTransition(to: target) else {
            throw .invalidLifecycleTransition(from: current, to: target)
        }
        guard effectiveAt <= recordedAt else {
            throw .transitionInFuture
        }
        // The start date bounds when a sankalpa may be *acted on*, so it gates entering In
        // progress. Ending one early is always allowed: a sankalpa declared for next week can be
        // stopped today, and simply has no evaluated periods.
        guard target.isTerminal || effectiveAt.day >= commitmentStart else {
            throw .transitionBeforeStart(startDate: commitmentStart)
        }
        // S14 — only the first transition, Not started → In progress, may separate its effective
        // time from when it was recorded. Everything else takes effect when the user performs it.
        guard effectiveAt == recordedAt || (current == .notStarted && target == .inProgress) else {
            throw .invalidLifecycleTransition(from: current, to: target)
        }
        // A change must land after everything already recorded. For a backdated Begin that is
        // the user's choice to correct; for every other transition it means the device clock now
        // reads earlier than it did — travelling west, or the hour daylight saving gives back —
        // so it is reported as the clock problem it is rather than as an illegal transition.
        if let last = transitions.last, effectiveAt < last.effectiveAt {
            throw .transitionOutOfOrder(lastRecorded: last.effectiveAt)
        }

        transitions.append(
            LifecycleTransition(
                from: current,
                to: target,
                effectiveAt: effectiveAt,
                recordedAt: recordedAt
            )
        )
        current = target
    }

    // MARK: - Reconstruction

    /// The state the sankalpa was in at `moment`, from the effective times.
    public func state(at moment: CalendarMoment) -> LifecycleState {
        var state = LifecycleState.notStarted
        for transition in transitions where transition.effectiveAt <= moment {
            state = transition.to
        }
        return state
    }

    /// S10 — the lifecycle half of session eligibility.
    public func wasInProgress(at moment: CalendarMoment) -> Bool {
        state(at: moment) == .inProgress
    }

    /// S13 — true only when the sankalpa was Paused for the window's entire span. A window with any
    /// In progress time in it is evaluated normally (DD-14).
    public func wasPausedThroughout(_ window: PeriodWindow) -> Bool {
        guard state(at: window.firstMoment) == .paused else { return false }
        return !transitions.contains {
            $0.effectiveAt > window.firstMoment && $0.effectiveAt <= window.lastMoment
        }
    }

    /// The transition into a terminal state, if the user has made one. It is the reporting cutoff
    /// for period outcomes (DD-16).
    public var terminalTransition: LifecycleTransition? {
        transitions.last { $0.to.isTerminal }
    }

    public var beganAt: CalendarMoment? {
        transitions.first { $0.to == .inProgress }?.effectiveAt
    }

    public func canTransition(to target: LifecycleState) -> Bool {
        current.canTransition(to: target)
    }
}

File: backend/src/main/java/com/sankalpa/domain/lifecycle/LifecycleTimeline.java
package com.sankalpa.domain.lifecycle;

import com.sankalpa.domain.DomainException;
import com.sankalpa.domain.commitment.PeriodWindow;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

public final class LifecycleTimeline {
    private LifecycleState current;
    private final List<LifecycleTransition> transitions;

    public LifecycleTimeline() {
        this(LifecycleState.NOT_STARTED, List.of());
    }

    public LifecycleTimeline(LifecycleState current, List<LifecycleTransition> transitions) {
        this.current = current;
        this.transitions = new ArrayList<>(transitions);
    }

    public LifecycleState current() { return current; }
    public List<LifecycleTransition> transitions() { return List.copyOf(transitions); }

    public void begin(LocalDate startDate, LocalDateTime effectiveAt, LocalDateTime recordedAt) {
        if (effectiveAt.isAfter(recordedAt)) {
            throw new DomainException("TRANSITION_IN_FUTURE", "Begin cannot be effective in the future");
        }
        if (effectiveAt.toLocalDate().isBefore(startDate)) {
            throw new DomainException("TRANSITION_BEFORE_START", "Begin cannot be effective before the start date");
        }
        transitionTo(LifecycleState.IN_PROGRESS, effectiveAt, recordedAt);
    }

    public void transitionNow(LifecycleState target, LocalDateTime now) {
        transitionTo(target, now, now);
    }

    private void transitionTo(LifecycleState target, LocalDateTime effectiveAt, LocalDateTime recordedAt) {
        if (!current.canTransitionTo(target)) {
            throw new DomainException("INVALID_LIFECYCLE_TRANSITION",
                    "Cannot transition from " + current + " to " + target);
        }
        transitions.add(new LifecycleTransition(current, target, effectiveAt, recordedAt));
        current = target;
    }

    public LifecycleState stateAt(LocalDateTime at) {
        LifecycleState state = LifecycleState.NOT_STARTED;
        List<LifecycleTransition> ordered = transitions.stream()
                .sorted(Comparator.comparing(LifecycleTransition::effectiveAt))
                .toList();
        for (LifecycleTransition transition : ordered) {
            if (transition.effectiveAt().isAfter(at)) break;
            state = transition.to();
        }
        return state;
    }

    public boolean wasInProgressAt(LocalDateTime at) {
        return stateAt(at) == LifecycleState.IN_PROGRESS;
    }

    public boolean wasPausedThroughout(PeriodWindow window) {
        LocalDateTime start = window.start().atStartOfDay();
        LocalDateTime end = window.end().atTime(LocalTime.MAX);
        if (stateAt(start) != LifecycleState.PAUSED) return false;
        return transitions.stream()
                .noneMatch(t -> t.effectiveAt().isAfter(start)
                        && !t.effectiveAt().isAfter(end)
                        && t.to() != LifecycleState.PAUSED);
    }

    public Optional<LocalDateTime> terminalAt() {
        return transitions.stream()
                .filter(t -> t.to().terminal())
                .map(LifecycleTransition::effectiveAt)
                .findFirst();
    }
}
